checker/jdk/nullness/src/java/beans/PropertyDescriptor.java

```diff
@@ -0,0 +1,754 @@
+            if (isAssignable(xr, yr)) {
+                setReadMethod(yr);
+            } else {
+                setReadMethod(xr);
+            }
+        } catch (IntrospectionException ex) {
+            // fall through
+        }
+
+        // However, if both x and y reference read methods in the same class,
+        // give priority to a boolean "is" method over a boolean "get" method.
+        if (xr != null && yr != null &&
+                   xr.getDeclaringClass() == yr.getDeclaringClass() &&
+                   getReturnType(getClass0(), xr) == boolean.class &&
+                   getReturnType(getClass0(), yr) == boolean.class &&
+                   xr.getName().indexOf(Introspector.IS_PREFIX) == 0 &&
+                   yr.getName().indexOf(Introspector.GET_PREFIX) == 0) {
+            try {
+                setReadMethod(xr);
+            } catch (IntrospectionException ex) {
+                // fall through
+            }
+        }
+
+        Method xw = x.getWriteMethod();
+        Method yw = y.getWriteMethod();
+
+        try {
+            if (yw != null) {
+                setWriteMethod(yw);
+            } else {
+                setWriteMethod(xw);
+            }
+        } catch (IntrospectionException ex) {
+            // Fall through
+        }
+
+        if (y.getPropertyEditorClass() != null) {
+            setPropertyEditorClass(y.getPropertyEditorClass());
+        } else {
+            setPropertyEditorClass(x.getPropertyEditorClass());
+        }
+
+
+        bound = x.bound | y.bound;
+        constrained = x.constrained | y.constrained;
+    }
+
+    /*
+     * Package-private dup constructor.
+     * This must isolate the new object from any changes to the old object.
+     */
+    PropertyDescriptor(PropertyDescriptor old) {
+        super(old);
+        propertyTypeRef = old.propertyTypeRef;
+        this.readMethodRef.set(old.readMethodRef.get());
+        this.writeMethodRef.set(old.writeMethodRef.get());
+        propertyEditorClassRef = old.propertyEditorClassRef;
+
+        writeMethodName = old.writeMethodName;
+        readMethodName = old.readMethodName;
+        baseName = old.baseName;
+
+        bound = old.bound;
+        constrained = old.constrained;
+    }
+
+    void updateGenericsFor(Class<?> type) {
+        setClass0(type);
+        try {
+            setPropertyType(findPropertyType(this.readMethodRef.get(), this.writeMethodRef.get()));
+        }
+        catch (IntrospectionException exception) {
+            setPropertyType(null);
+        }
+    }
+
+    /**
+     * Returns the property type that corresponds to the read and write method.
+     * The type precedence is given to the readMethod.
+     *
+     * @return the type of the property descriptor or null if both
+     *         read and write methods are null.
+     * @throws IntrospectionException if the read or write method is invalid
+     */
+    private Class<?> findPropertyType(Method readMethod, Method writeMethod)
```

Comment:
Per the documentation, I would mark both parameters and the return type as `@Nullable`, even though the method is private.